httpclient/api_client.go

```diff
@@ -0,0 +1,386 @@
+		}
+		for _, requestVisitor := range visitors {
+			err = requestVisitor(request)
+			if err != nil {
+				return c.failRequest(ctx, "failed during request visitor", err)
+			}
+		}
+		// request.Context() holds context potentially enhanced by visitors
+		request.Header.Set("User-Agent", useragent.FromContext(request.Context()))
+
+		// attempt the actual request
+		response, err := c.httpClient.Do(request)
+
+		// After this point, the request body has (probably) been consumed. handleError() must be called to reset it if
+		// possible.
+		if _, ok := err.(*url.Error); ok {
+			return c.handleError(ctx, err, requestBody)
+		}
+
+		// By this point, the request body has certainly been consumed.
+		responseBody, responseBodyErr := c.fromResponse(response)
+		if responseBodyErr != nil {
+			return c.failRequest(ctx, "failed while reading response", responseBodyErr)
+		}
+
+		mappedError := c.config.ErrorMapper(ctx, response, responseBody.ReadCloser)
+		defer c.recordRequestLog(ctx, request, response, mappedError, requestBody.DebugBytes, responseBody.DebugBytes)
+
+		if mappedError == nil {
+			return &responseBody, nil
+		}
+
+		// proactively release the connections in HTTP connection pool
+		c.httpClient.CloseIdleConnections()
+		return c.handleError(ctx, mappedError, requestBody)
+	}
+}
+
+func (c *ApiClient) recordRequestLog(
+	ctx context.Context,
+	request *http.Request,
+	response *http.Response,
+	err error,
+	requestBody, responseBody []byte,
+) {
+	// Don't compute expensive debug message if debug logging is not enabled.
+	if !logger.Get(ctx).Enabled(ctx, logger.LevelDebug) {
+		return
+	}
+	sb := strings.Builder{}
+	sb.WriteString(fmt.Sprintf("%s %s", request.Method,
+		escapeNewLines(request.URL.Path)))
+	if request.URL.RawQuery != "" {
+		sb.WriteString("?")
+		q, _ := url.QueryUnescape(request.URL.RawQuery)
+		sb.WriteString(q)
+	}
+	sb.WriteString("\n")
+	if c.config.DebugHeaders {
+		sb.WriteString("> * Host: ")
+		sb.WriteString(escapeNewLines(request.Host))
+		sb.WriteString("\n")
+		for k, v := range request.Header {
+			trunc := onlyNBytes(strings.Join(v, ""), c.config.DebugTruncateBytes)
+			sb.WriteString(fmt.Sprintf("> * %s: %s\n", k, escapeNewLines(trunc)))
+		}
+	}
+	if len(requestBody) > 0 {
+		sb.WriteString(c.redactedDump("> ", requestBody))
+		sb.WriteString("\n")
+	}
+	sb.WriteString("< ")
+	if response != nil {
+		sb.WriteString(fmt.Sprintf("%s %s", response.Proto, response.Status))
+		// Only display error on this line if the response body is empty.
+		// Otherwise the response body will include details about the error.
+		if len(responseBody) == 0 && err != nil {
+			sb.WriteString(fmt.Sprintf(" (Error: %s)", err))
+		}
+	} else {
+		sb.WriteString(fmt.Sprintf("Error: %s", err))
+	}
+	sb.WriteString("\n")
+	if len(responseBody) > 0 {
+		sb.WriteString(c.redactedDump("< ", responseBody))
+	}
+	logger.Debugf(ctx, sb.String()) // lgtm [go/log-injection] lgtm [go/clear-text-logging]
```

Comment:
we should remove go links since this is public repo.

Reply:
@tanmay-db those are not go links - those are linter rules